File: dotnet/src/DocGen/Clients/PdfToolsClient.cs
using Dokmatiq.DocGen.Internal;

namespace Dokmatiq.DocGen.Clients;

/// <summary>
/// Client for PDF manipulation tools.
/// </summary>
public sealed class PdfToolsClient
{
    private readonly Transport _transport;

    internal PdfToolsClient(Transport transport) => _transport = transport;

    /// <summary>Merge multiple PDF files into one.</summary>
    public byte[] Merge(IEnumerable<string> filePaths)
    {
        var pdfs = filePaths.Select(FileUtils.ToBase64).ToList();
        return _transport.RequestBytes(HttpMethod.Post, "/api/pdf-tools/merge", new { pdfs });
    }

    /// <summary>Merge PDFs from byte arrays.</summary>
    public byte[] Merge(params byte[][] pdfs)
    {
        var encoded = pdfs.Select(FileUtils.ToBase64).ToList();
        return _transport.RequestBytes(HttpMethod.Post, "/api/pdf-tools/merge", new { pdfs = encoded });
    }

    /// <summary>Split a PDF by page ranges.</summary>
    public byte[] Split(string filePath, string pages)
        => _transport.RequestBytes(HttpMethod.Post, "/api/pdf-tools/split",
            new { pdfBase64 = FileUtils.ToBase64(filePath), pages });

    /// <summary>Extract text from a PDF.</summary>
    public string ExtractText(string filePath)
    {
        var result = _transport.RequestJson<Dictionary<string, object>>(HttpMethod.Post,
            "/api/pdf-tools/extract-text", new { pdfBase64 = FileUtils.ToBase64(filePath) });
        return result.TryGetValue("text", out var text) ? text?.ToString() ?? "" : "";
    }

    /// <summary>Extract text from bytes.</summary>
    public string ExtractText(byte[] data)
    {
        var result = _transport.RequestJson<Dictionary<string, object>>(HttpMethod.Post,
            "/api/pdf-tools/extract-text", new { pdfBase64 = FileUtils.ToBase64(data) });
        return result.TryGetValue("text", out var text) ? text?.ToString() ?? "" : "";
    }

    /// <summary>Get PDF metadata.</summary>
    public Dictionary<string, object> GetMetadata(string filePath)
        => _transport.RequestJson<Dictionary<string, object>>(HttpMethod.Post,
            "/api/pdf-tools/metadata", new { pdfBase64 = FileUtils.ToBase64(filePath) });

    /// <summary>Set PDF metadata.</summary>
    public byte[] SetMetadata(string filePath, Dictionary<string, string> metadata)
    {
        var body = new Dictionary<string, object> { ["pdfBase64"] = FileUtils.ToBase64(filePath) };
        foreach (var (key, value) in metadata) body[key] = value;
        return _transport.RequestBytes(HttpMethod.Post, "/api/pdf-tools/metadata/set", body);
    }

    /// <summary>Convert to PDF/A.</summary>
    public byte[] ToPdfA(string filePath)
        => _transport.RequestBytes(HttpMethod.Post, "/api/pdf-tools/pdfa",
            new { pdfBase64 = FileUtils.ToBase64(filePath) });

    /// <summary>Convert bytes to PDF/A.</summary>
    public byte[] ToPdfA(byte[] data)
        => _transport.RequestBytes(HttpMethod.Post, "/api/pdf-tools/pdfa",
            new { pdfBase64 = FileUtils.ToBase64(data) });

    /// <summary>Rotate pages in a PDF.</summary>
    public byte[] Rotate(string filePath, int angle, string? pages = null)
    {
        var body = new Dictionary<string, object>
        {
            ["pdfBase64"] = FileUtils.ToBase64(filePath),
            ["angle"] = angle
        };
        if (pages is not null) body["pages"] = pages;
        return _transport.RequestBytes(HttpMethod.Post, "/api/pdf-tools/rotate", body);
    }
}

File: java/src/main/java/com/dokmatiq/docgen/client/PdfToolsClient.java
package com.dokmatiq.docgen.client;

import com.dokmatiq.docgen.internal.FileUtils;
import com.dokmatiq.docgen.internal.Transport;

import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/** Client for PDF manipulation tools. */
public class PdfToolsClient {
    private final Transport transport;

    public PdfToolsClient(Transport transport) {
        this.transport = transport;
    }

    /** Merge multiple PDFs into one. */
    public byte[] merge(List<Path> files) {
        var pdfs = files.stream().map(FileUtils::toBase64).collect(Collectors.toList());
        return transport.requestBytes("POST", "/api/pdf-tools/merge", Map.of("pdfs", pdfs));
    }

    /** Merge PDFs from byte arrays. */
    public byte[] merge(byte[]... pdfs) {
        var encoded = java.util.Arrays.stream(pdfs).map(FileUtils::toBase64).collect(Collectors.toList());
        return transport.requestBytes("POST", "/api/pdf-tools/merge", Map.of("pdfs", encoded));
    }

    /** Extract text content from a PDF. */
    public String extractText(Path file) {
        var result = transport.requestJson("POST", "/api/pdf-tools/extract-text",
                Map.of("pdfBase64", FileUtils.toBase64(file)), Map.class);
        return String.valueOf(result.get("text"));
    }

    /** Extract text from bytes. */
    public String extractText(byte[] data) {
        var result = transport.requestJson("POST", "/api/pdf-tools/extract-text",
                Map.of("pdfBase64", FileUtils.toBase64(data)), Map.class);
        return String.valueOf(result.get("text"));
    }

    /** Get PDF metadata. */
    public Map<?, ?> getMetadata(Path file) {
        return transport.requestJson("POST", "/api/pdf-tools/metadata",
                Map.of("pdfBase64", FileUtils.toBase64(file)), Map.class);
    }

    /** Set PDF metadata. */
    public byte[] setMetadata(Path file, Map<String, String> metadata) {
        var body = new HashMap<String, Object>();
        body.put("pdfBase64", FileUtils.toBase64(file));
        body.putAll(metadata);
        return transport.requestBytes("POST", "/api/pdf-tools/metadata/set", body);
    }

    /** Convert to PDF/A. */
    public byte[] toPdfA(Path file) {
        return transport.requestBytes("POST", "/api/pdf-tools/pdfa",
                Map.of("pdfBase64", FileUtils.toBase64(file)));
    }

    /** Convert bytes to PDF/A. */
    public byte[] toPdfA(byte[] data) {
        return transport.requestBytes("POST", "/api/pdf-tools/pdfa",
                Map.of("pdfBase64", FileUtils.toBase64(data)));
    }

    /** Rotate pages in a PDF. */
    public byte[] rotate(Path file, int angle) {
        return transport.requestBytes("POST", "/api/pdf-tools/rotate",
                Map.of("pdfBase64", FileUtils.toBase64(file), "angle", angle));
    }

    /** Rotate specific pages. */
    public byte[] rotate(Path file, int angle, String pages) {
        var body = new HashMap<String, Object>();
        body.put("pdfBase64", FileUtils.toBase64(file));
        body.put("angle", angle);
        body.put("pages", pages);
        return transport.requestBytes("POST", "/api/pdf-tools/rotate", body);
    }
}
